Feature/#7 recipe Create, Read 로직 구현

buchonsi opened the pull request:
레시피 등록, 조회 기능 구현

* 기본적인 요구사항에 필요한 service 로직 구현
* recipe 필드에 추가되어야 하는 값이 필요 -> 추후 구현

buchonsi committed:
[#7] 레시피 등록, 조회 단순 기능 추가

기본 동작 기능인 생성, id로 조회, title로 조회 기능을 추가 후 커밋

buchonsi committed:
[#7] Recipe 도메인 RecipeType 필드 추가

* 클라이언트에서 RecipeType 배열이 넘어오면 DB 에 문자열로 저장됨
* RecipeType 필드 추가에 따른 테스트 코드 수정

buchonsi committed:
[#7] RecipeController 추가

* 저장, ID로 recipe 찾기, title로 recipe 찾기 구현

buchonsi committed:
[#7] 주석 정리

f-lab-maru commented on src/main/java/com/flab/recipebook/recipe/controller/RecipeController.java:
```diff
@@ -0,0 +1,34 @@
+        return new ResponseEntity<>(new ResponseResult(recipeService.findById(recipeId)), HttpStatus.OK);
+    }
+
+    @GetMapping("/recipe/search/{title}")
```
지금은 타이틀로만 검색을 하고 있는데 다른 필드로도 검색할 경우가 생길것 같은데요.
이 부분은 확장성이 부족한거 같습니다

buchonsi replied:
확장성을 고려해 타입과 키워드를 가진 객체를 받아 조건 검색을 하도록 코드를 수정하였습니다.

https://github.com/f-lab-edu/RecipeBook/blob/8ed7b2c1fa5c8a3c2e5f69311ced03fab4809381/src/main/java/com/flab/recipebook/recipe/controller/RecipeController.java#L31-L34  


https://github.com/f-lab-edu/RecipeBook/blob/8ed7b2c1fa5c8a3c2e5f69311ced03fab4809381/src/main/java/com/flab/recipebook/recipe/dto/SearchRecipeDto.java#L3-L5  


https://github.com/f-lab-edu/RecipeBook/blob/8ed7b2c1fa5c8a3c2e5f69311ced03fab4809381/src/main/resources/mapper/RecipeMapper.xml#L99-L106

f-lab-maru commented on src/main/java/com/flab/recipebook/recipe/domain/dao/RecipeDao.java:
```diff
@@ -0,0 +1,18 @@
+public interface RecipeDao {
+    void save(Recipe recipe);
+
+    Optional<Recipe> findById(Long id);
```
Optional 로 하신 이유가 있나요?

buchonsi replied:
Optional 을 이용하면 null 처리를 조금 더 편하게 할 수 있고 코드가 간결해지기 때문에 사용을 하였습니다. 
또한 나중에 DB를 JPA 를 사용하여 접근하도록 변경할 때 서비스 코드의 수정을 줄이기 위해 Optional을 사용하였습니다.

f-lab-maru commented on src/main/java/com/flab/recipebook/recipe/dto/ResponseRecipeDto.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    public List<String> setRecipeType(String recipeType) {
+        if (recipeType == null) {
```
List<String> 이 아니라 스트링을 받아서 스플릿 하는 이유가 있나요?

buchonsi replied:
RecipeType 은 클라이언트에서 체크박스로 선택되어 Recipe 한 개에 여러개가 종속되어 저장이 되어야 했습니다. 

그래서 처음에 의도한 것은 RecipeType 은 개수가 정해져 있고 그 수가 많지 않기 때문에 선택되어진 RecipeType 을 DB에 그대로 저장을 하는 것이 좋다고 생각하였습니다. 

그 이유는 RecipeTpye 을 DB 정규화 하여 FK 로 사용하면 DB 호출이 추가로 필요하게 되고 그렇게 되면 성능에 영향을 준다고 생각하였기 때문입니다.

하지만 추가로 DB 호출을 하더라고 인덱스를 통해서 RecipeType을 가져오기 떄문에 성능에 큰 영향이 있을까 라는 고민도 되어서 어느 방법이 더 좋을지는 조금 더 고민해보아야 할 것 같습니다.

f-lab-maru commented on src/main/java/com/flab/recipebook/recipe/dto/SaveRecipeDto.java:
```diff
@@ -0,0 +1,42 @@
+        return recipeType;
+    }
+
+    public void setRecipeType(List<String> recipeType) {
```
setter 보다는 생성자로 받아서 생성자 내부에서 처리하는게 좋을거 같습니다

buchonsi replied:
상태값을 변경시키는 setter는 위험성이 있으므로 생성자 내부에서 처리하도록 변경하였습니다.

https://github.com/f-lab-edu/RecipeBook/blob/cdc7c73de6c6966a0d7d43c729f75fedc15d9975/src/main/java/com/flab/recipebook/recipe/dto/SaveRecipeDto.java#L13-L18

https://github.com/f-lab-edu/RecipeBook/blob/cdc7c73de6c6966a0d7d43c729f75fedc15d9975/src/main/java/com/flab/recipebook/recipe/dto/SaveRecipeDto.java#L36-L41

f-lab-maru commented on src/main/java/com/flab/recipebook/recipe/service/RecipeService.java:
```diff
@@ -0,0 +1,82 @@
+     * 레시피에는 제목, 재료, 내용을 작성할 수 있다.
+     */
+    public void saveRecipe(SaveRecipeDto saveRecipeDto) {
+        //권한 검증
```
권한 검증은 처리가 안된건가요?

buchonsi replied:
스프링 시큐리티 설정에서 url 에 따라 접근 권한을 추가하었고, 
레시피를 등록하는 것은 프로젝트 서비스에서 중요한 기능이라고 생각되며, 그래서 service에서 한번 더 권한을 검증 하는 로직을 추가하였습니다.

https://github.com/f-lab-edu/RecipeBook/blob/cdc7c73de6c6966a0d7d43c729f75fedc15d9975/src/main/java/com/flab/recipebook/common/config/security/SecurityConfiguration.java#L54-L56

https://github.com/f-lab-edu/RecipeBook/blob/cdc7c73de6c6966a0d7d43c729f75fedc15d9975/src/main/java/com/flab/recipebook/recipe/service/RecipeService.java#L82-L88

f-lab-maru commented on src/main/java/com/flab/recipebook/recipe/service/RecipeService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public ResponseRecipeDto convertResponseRecipe(Recipe recipe) {
+        User user = userDao.findById(recipe.getUserNo());
```
recipeDao.findById 는 optional 인데 userDao.findById 는 optional 이 아닌 이유가 있나요?

buchonsi replied:
초반에 User 서비스를 만들 때 Optional 을 생각하지 못해서 누락된 것 같습니다.
코드가 통일 될 수 있도록 findById의 리턴 값도 Optional으로 수정하였습니다.

f-lab-maru commented on src/main/resources/mapper/RecipeMapper.xml:
```diff
@@ -0,0 +1,53 @@
+            create_date,
+            modify_date
+        from Recipe
+        WHERE title like "%" || #{title} || "%"
```
앞뒤 like 쿼리가 걸려있는데요. 
이 쿼리는 일반적으로 인덱스를 타지 않는데, 그 이유를 아실까요?

buchonsi replied:
* like 쿼리에 % 를 앞쪽에 사용하게 되면 정렬의 순서를 사용할 수 없어서 Full Scan 으로 데이터가 조회가 되고, 인덱스를 타지 못하므로 조회 속도가 느려집니다.

* 속도를 개선하는 방법으로 Mysql 에서는 Full text index 를 사용하여 속도를 개선 할 수 있습니다.

* Full text index 로 검색 방법을 프로젝트에 적용하면서 문제점이나 다른 단점이 없는지 조금 더 고민해 보아야 할 것 같습니다.

buchonsi committed:
Merge branch 'master' into feature/#7-recipe-CR

buchonsi committed:
[#7] 확장성을 고려하여 레시피 검색 로직 변경

- 검색 키워드를 한가지 필드가 아닌 다양한 필드로 받을 수 있도록 클래스 추가.
- DB에서 recipe를 가져오는 방법을 sql join 으로 변경 -> sevice에서 처리하려면 db 호출이 너무 많아지기 때문에

buchonsi committed:
[#7] Recipe 전체 조회 로직 추가

buchonsi committed:
[#11] 유저 조회시 return 타입 Optional로 변경

- return 타입을 이용하여 null처리와 exception 처리 등 편리한 기능과 코드를 조금 더 깔끔하게 하기 위해 리턴 값을 Optional로 변경
- DB 접근을 JPA로 변경하여도 서비스 코드의 수정이 많이 없도록 하기 위해 Optional로 변경하는 것이 좋다고 판단.

buchonsi committed:
[#7] 예외 처리 반복되는 코드 리팩토링

buchonsi committed:
[#7] 도메인 별 접근 권한 설정 및 테스트

- 권한 별로 도메인을 접근하도록 스프링 시큐리티 설정 변경

buchonsi committed:
[#7] Setter -> 생성자 내부에서 처리하도록 변경

- 상태값을 변환시키는 setter는 위험할 수 있으므로 생성자 내부에서 처리
